Client/Frontend/Shields/BraveShield.swift

```diff
@@ -0,0 +1,145 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveShared
+import Data
+import CoreData
+
+// These override the setting in the prefs
+public enum BraveShield {
+    case AllOff
+    case AdblockAndTp
+    case HTTPSE
+    case SafeBrowsing
+    case FpProtection
+    case NoScript
+    
+    var globalPreference: Bool {
+        switch self {
+        case .AllOff:
+            return false
+        case .AdblockAndTp:
+            return Preferences.Shields.blockAdsAndTracking.value
+        case .HTTPSE:
+            return Preferences.Shields.httpsEverywhere.value
+        case .SafeBrowsing:
+            return Preferences.Shields.blockPhishingAndMalware.value
+        case .FpProtection:
+            return Preferences.Shields.fingerprintingProtection.value
+        case .NoScript:
+            return Preferences.Shields.blockScripts.value
+        }
+    }
+}
+
+private class PrivateBrowsingShieldOverride {
+    fileprivate static var privateModeOverrides: [String: PrivateBrowsingShieldOverride] = [:]
+    
+    private var overrides: [BraveShield: Bool] = [:]
+    
+    init(shield: BraveShield, isOn: Bool) {
+        overrides[shield] = isOn
+    }
+    
+    subscript(shield: BraveShield) -> Bool? {
+        get { return overrides[shield] }
+        set { overrides[shield] = newValue }
+    }
+}
+
+extension Domain {
+    
+    /// Remove all private browsing shield overrides
+    class func resetPrivateBrowsingShieldOverrides() {
+        PrivateBrowsingShieldOverride.privateModeOverrides.removeAll()
+    }
+    
+    class func setBraveShield(forUrl url: URL, shield: BraveShield, isOn: Bool?,
+                              context: NSManagedObjectContext = DataController.newBackgroundContext()) {
+        
+        let domain = Domain.getOrCreateForUrl(url, context: context)
```

Comment:
One thing that we should take care of here, is not store privately visited domains in CD unless user explicitly requests this. This would be like saving a favorite or bookmark. 

Inside here, we should check against `isPrivate`, and if it is, then save to array and completely bypass any concept of `Domain`.